### DISCUSSION TITLE
Alternative to DTE for Debugger and calling built-in VS commands?

### DISCUSSION THREAD

**Discussion by moon6969** [VsixCommunity/Community.VisualStudio.Toolkit#299] [Q&A]
I read discussion [#66 Should DTE be avoided?](https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/discussions/66) with interest.

My Google-Fu is failing me trying to find non-DTE methods for my Extension.
Are there any migration guides, references etc. about?
(I've seen the [VSIX Cookbook](https://www.vsixcookbook.com) that has some recipes & snippets.)

Is there a non-DTE way for the following? (don't need code, just hints on where to look!).
- iterate, create & edit Breakpoints (dte.Debugger.Breakpoints...)
- use an UndoContext (dte.UndoContext...)
- execute a built-in VS command (dte.Commands.Raise(formatSelectionCommand.Guid, ...))

Thanks.

**Comment by ErikEJ** [comment]
You can still use DTE if your extension targets only VS 2022

**Reply by moon6969** [reply]
Thanks Eric, my VS2022 extension does currently work as desired using DTE2.

I'm not sure why you say "if your extension targets only VS 2022"? What is the wrinkle if I wanted to target say VS2019 as well?

**Reply by ErikEJ** [reply]
Then you would need to build and publish two distinct extensions, (maybe from as shared codebase), as the "old" DTE references in VS 2019 are not allowed in VS 2022.

**Reply by moon6969** [reply]
Found one answer in the Object Browser...
`await VS.Commands.ExecuteAsync(KnownCommands.Edit_FormatSelection);`

**Comment by reduckted** [comment]
Replacement for the UndoContext: https://github.com/VsixCommunity/Community.VisualStudio.Toolkit/blob/e2d88dd215a6453c108b1747c2f9b76f6e547063/src/toolkit/Community.VisualStudio.Toolkit.Shared/ExtensionMethods/ITextBufferExtensions.cs#L19
